### PR TITLE
[App] Add status endpoint, enable `ready`

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

This PR adds a status endpoint to the app API (via the api publish queue). The status endpoint communicates:
- whether the UI is ready
- the status of the works

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: examples/app_boring/app.py [hunk 1]
```diff
@@ -43,6 +43,10 @@ def __init__(self):
             raise_exception=True,
         )
 
+    @property
+    def ready(self) -> bool:
+        return self.source_work.has_succeeded and self.dest_work.is_running
```
Comment (tchaton) [inline examples/app_boring/app.py:48]: The second one is conditioned on the first one to succeeded already in the flow.

```suggestion
        return self.dest_work.is_running
```

File: src/lightning_app/core/api.py [hunk 2]
```diff
@@ -114,6 +123,13 @@ def run_once(self):
         except queue.Empty:
             pass
 
+        if self.app_status_queue is not None:
```
Comment (tchaton) [inline src/lightning_app/core/api.py:126]: I don't think this is required.
We are already pulling the state above. Let's simply perform status comparaison there operation there.

File: src/lightning_app/core/queues.py [hunk 3]
```diff
@@ -96,6 +97,10 @@ def get_api_delta_queue(self, queue_id: Optional[str] = None) -> "BaseQueue":
         queue_name = f"{queue_id}_{DELTA_QUEUE_CONSTANT}" if queue_id else DELTA_QUEUE_CONSTANT
         return self.get_queue(queue_name)
 
+    def get_app_status_queue(self, queue_id: Optional[str] = None) -> "BaseQueue":
```
Comment (tchaton) [inline src/lightning_app/core/queues.py:100]: Let's not add another queue, this is hard to handle properly in the cloud with the launcher.

File: src/lightning_app/core/app.py [hunk 4]
```diff
@@ -506,7 +503,7 @@ def _run(self) -> bool:
 
             self._update_run_time_monitor()
 
-            if self.ready and self._has_updated and self.should_publish_changes_to_api and self.api_publish_state_queue:
+            if self._has_updated and self.should_publish_changes_to_api and self.api_publish_state_queue:
                 self.api_publish_state_queue.put(self.state_vars)
```
Comment (tchaton) [inline src/lightning_app/core/app.py:507]: Alternatively, let's pass a tuple there (state, status) even if status can be inferred from state.

Comment (ethanwharris) [reply src/lightning_app/core/app.py]: Went for the tuple approach to avoid adding too much into state

**Review by awaelchli — APPROVED** [review]

File: src/lightning_app/core/app.py [hunk 5]
```diff
@@ -532,6 +529,21 @@ def _update_is_headless(self) -> None:
...
+        work_statuses = []
+        for work in breadth_first(self.root, types=(lightning_app.LightningWork,)):
+            work_statuses.append(work.status)
+
+        self.status = AppStatus(
+            is_ui_ready=self.ready,
+            work_statuses=work_statuses,
+        )
+
+        if self.status != old_status:
```
Comment (tchaton) [inline src/lightning_app/core/app.py:544]: This shouldn't be there. Only delta state can activate this.

**Review by tchaton — APPROVED** [review]

File: examples/app_boring/app.py [hunk 6]
```diff
@@ -43,6 +43,10 @@ def __init__(self):
             raise_exception=True,
         )
 
+    @property
+    def ready(self) -> bool:
```
Comment (Borda) [inline examples/app_boring/app.py:47]: shall this be property for the base class?

File: tests/tests_app/core/test_lightning_api.py [hunk 7]
```diff
@@ -195,9 +196,9 @@ def test_update_publish_state_and_maybe_refresh_ui():
     publish_state_queue = _MockQueue("publish_state_queue")
     api_response_queue = _MockQueue("api_response_queue")
 
-    publish_state_queue.put(app.state_with_changes)
+    publish_state_queue.put((app.state_with_changes, None))
 
-    thread = UIRefresher(publish_state_queue, api_response_queue)
+    thread = UIRefresher(publish_state_queue, api_response_queue, None)
```
Comment (awaelchli) [inline tests/tests_app/core/test_lightning_api.py:201]: perhaps we could use named arguments here for clarity.

**Review by awaelchli — APPROVED** [review]

File: examples/app_boring/app.py (hunk 6 above)
Comment (awaelchli) [reply examples/app_boring/app.py:47]: It is already part of the LightningFlow base class

File: tests/tests_app/core/test_lightning_api.py (hunk 7 above)
Comment (ethanwharris) [reply tests/tests_app/core/test_lightning_api.py]: Whoops, that was there by accident, removed :smiley:

**Pull request merged**